File: 0994-rotting-oranges/0994-rotting-oranges.cpp
class Solution {
public:
    int orangesRotting(vector<vector<int>>& grid) {
        queue<pair<int,int>> q;
        int freshOrange = 0;
        for(int i=0;i<grid.size();++i)
        {
            for(int j=0;j<grid[0].size();++j)
            {
                if(grid[i][j] == 2)
                    q.push({i,j});
                if(grid[i][j] == 1)
                    freshOrange++;
            }
        }
        
        int count = 0;
        while(!q.empty())
        {
            int sz = q.size();
            while(sz--)
            {
                auto i = q.front().first;
                auto j = q.front().second;
                q.pop();
                int drow[]={0,1,0,-1};
                int dcol[]={1,0,-1,0};
                for(int x=0;x<4;++x)
                {
                    int nrow = i+drow[x];
                    int ncol = j+dcol[x];
                    if(nrow>=0&&nrow<grid.size()&&ncol>=0&&ncol<grid[0].size()&&grid[nrow][ncol]==1)
                    {
                        grid[nrow][ncol]=2;
                        q.push({nrow,ncol});
                    }
                }
            }
            count++;
        }
        for(int i=0;i<grid.size();++i)
        {
            for(int j=0;j<grid[0].size();++j)
                if(grid[i][j] == 1)
                    return -1;
        }
        if(freshOrange > 0)
            return count-1;
        return 0;
    }
};

File: 0994-rotting-oranges/0994-rotting-oranges.java
class Solution {
    class Pair{
        int r;
        int c;
        int time;
        Pair(int r, int c, int time){
            this.r = r;
            this.c = c;
            this.time = time;
        }
    }
    public int orangesRotting(int[][] grid) {
        Queue<Pair> que = new ArrayDeque<>();
        int count = 0;
        for(int i = 0; i < grid.length; i++){
            for(int j = 0; j < grid[0].length; j++){
                if(grid[i][j] == 2){
                    que.add(new Pair(i, j, 0));
                }
                if(grid[i][j] == 1){
                    count++;
                }
            }
        }

        int[][] dir = {{-1, 0}, {0, -1}, {1,0}, {0, 1}};
        int time = 0;
        while(que.size() > 0){
            Pair curr = que.remove();
            time = Math.max(time, curr.time);
            for(int[] d: dir){
                int nr = curr.r + d[0];
                int nc = curr.c + d[1];

                if(nr < 0 || nc < 0 || nr >= grid.length || nc >= grid[0].length) continue;

                if(grid[nr][nc] == 1){
                    que.add(new Pair(nr, nc, curr.time + 1));
                    count--;
                    grid[nr][nc] = 2;
                }
            }
        }

        if(count > 0) return -1;

        return time;
    }
}